storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java

```diff
@@ -192,8 +195,13 @@ public void sanityCheck(boolean timeIndexFileNewlyCreated) throws IOException {
      * the time index).
      */
     public TimestampOffset readMaxTimestampAndOffsetSoFar() throws IOException {
-        if (maxTimestampAndOffsetSoFar == TimestampOffset.UNKNOWN)
-            maxTimestampAndOffsetSoFar = timeIndex().lastEntry();
+        if (maxTimestampAndOffsetSoFar == TimestampOffset.UNKNOWN) {
+            synchronized (maxTimestampAndOffsetLock) {
```

Comment:
Could you please add comments for this lock? By default, `LogSegment` is not thread-safe so all modification should be executed within `UnifiedLog` lock. However, there is an exceptional path that could change the `maxTimestampAndOffsetSoFar` without `UnifiedLog` lock, which causes this concurrent issue ...

Reply:
Sounds good. Added a comment